aws-java-sdk-sts/src/main/java/com/amazonaws/auth/STSSessionCredentialsProvider.java

```diff
@@ -116,39 +167,186 @@ public STSSessionCredentialsProvider(AWSCredentialsProvider longLivedCredentials
+            }
+        }
     }
 
+    /**
+     * Brute-force refresh of session credentials. A decision to use this method
+     * should be made judiciously since this class automatically manages
+     * refreshing expiring credentials limiting its usefulness. Calling this
+     * method may temporarily cause getCredentials() to block until a new session
+     * is fetched from the STS service.
+     */
     @Override
     public void refresh() {
-        startSession();
+        sessionCredentials.set(null);
+        blockingNewSession();
     }
-
+    
     /**
-     * Starts a new session by sending a request to the AWS Security Token
-     * Service (STS) with the long lived AWS credentials. This class then vends
-     * the short lived session credentials sent back from STS.
+     * Unconditionally calls STS to generate a new session
+     * 
+     * @return 
      */
-    private void startSession() {
+    private SessionCredentialsHolder newSession() {
         GetSessionTokenResult sessionTokenResult = securityTokenService
-                .getSessionToken(new GetSessionTokenRequest().withDurationSeconds(DEFAULT_DURATION_SECONDS));
+                .getSessionToken(new GetSessionTokenRequest()
+                        .withDurationSeconds(DEFAULT_DURATION_SECONDS));
+
         Credentials stsCredentials = sessionTokenResult.getCredentials();
 
-        sessionCredentials = new BasicSessionCredentials(
+        AWSSessionCredentials credentials = new BasicSessionCredentials(
                 stsCredentials.getAccessKeyId(),
                 stsCredentials.getSecretAccessKey(),
                 stsCredentials.getSessionToken());
-        sessionCredentialsExpiration = stsCredentials.getExpiration();
+
+        return new SessionCredentialsHolder(credentials, stsCredentials.getExpiration());
+    }
+    
+    /**
+     * Used when there is no valid session to return. Callers are blocked until
+     * a new session is created or an exception is thrown.
+     * Throws AmazonClientException if new session is not created
+     * within 5 seconds.
+     */
+    private void blockingNewSession() {
+        newSessionLock.lock();
+
+        try {
+            // return if successful refresh occured while waiting for the lock
+            if(!expired()) {
+                return;
+            }
+            
+            // trigger a refresh thread if needed
+            asyncNewSession();
+
+            // wait for new session or time out
+            if (false == newSessionCondition.await(NEW_SESSION_MAX_WAIT_SECONDS, TimeUnit.SECONDS)) {
+                // timeout occured, check for valid session before throwing an exception
+                if (expired()) {
+                    throw new AmazonClientException("Session credentials request timed-out.", lastException.get());
+                }
+            }
+        } catch (InterruptedException ex) {
+            throw new AmazonClientException(ex);
```

Comment:
I think calling interrupt on the thread here is not a bad idea since the apache http client does check the interrupted flag when it can. That could be a poor-man's solution to the transient socket timeout issue I was thinking about solving in a more deterministic fashion. Will do the former for now.